# 5. 

In Chapter 4, we used logistic regression to predict the probability of default using income and balance on the Default data set. We will now estimate the test error of this logistic regression model using the validation set approach. Do not forget to set a random seed before beginning your analysis.


In [1]:
library(ISLR)
data("Default")

Warning message:
"package 'ISLR' was built under R version 4.2.2"


## (a) 
Fit a logistic regression model that uses income and balance to predict default.

In [2]:
head(Default)

default student balance   income   
1 No      No       729.5265 44361.625
2 No      Yes      817.1804 12106.135
3 No      No      1073.5492 31767.139
4 No      No       529.2506 35704.494
5 No      No       785.6559 38463.496
6 No      Yes      919.5885  7491.559

In [3]:
Default$target <- ifelse(Default$default == "Yes", 1, 0)
Default$student_e <- ifelse(Default$student == "Yes", 1, 0)

In [4]:
head(Default)

default student balance   income    target student_e
1 No      No       729.5265 44361.625 0      0        
2 No      Yes      817.1804 12106.135 0      1        
3 No      No      1073.5492 31767.139 0      0        
4 No      No       529.2506 35704.494 0      0        
5 No      No       785.6559 38463.496 0      0        
6 No      Yes      919.5885  7491.559 0      1

In [5]:
options(scipen = 999)

In [6]:
glm.fit = glm(target~income+balance, data=Default, family=binomial)
summary(glm.fit)


Call:
glm(formula = target ~ income + balance, family = binomial, data = Default)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.4725  -0.1444  -0.0574  -0.0211   3.7245  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.540468437   0.434756357 -26.545 < 0.0000000000000002 ***
income        0.000020809   0.000004985   4.174            0.0000299 ***
balance       0.005647103   0.000227373  24.836 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2920.6  on 9999  degrees of freedom
Residual deviance: 1579.0  on 9997  degrees of freedom
AIC: 1585

Number of Fisher Scoring iterations: 8


## (b) 
Using the validation set approach, estimate the test error of this model. In order to do this, you must perform the following steps:
- i. Split the sample set into a training set and a validation set.
- ii. Fit a multiple logistic regression model using only the training observations.
- iii. Obtain a prediction of default status for each individual in the validation set by computing the posterior probability of default for that individual, and classifying the individual to the default category if the posterior probability is greater than 0.5.
- iv. Compute the validation set error, which is the fraction of the observations in the validation set that are misclassified.

In [7]:
set.seed (123)

train_test_split <- function(split=0.8){
    len = length(Default$target)
    train = sample(len, (len * split))
    
    X_train = Default[train,]
    X_test = Default[-train, ]
    y_test = Default$target[-train]
    
    return (list(
        'X_train'=X_train,
        'X_test'=X_test,
        'y_test'=y_test
    ))
}

eighty_twenty = train_test_split()

In [8]:
glm.b = glm(target~income+balance, data=eighty_twenty$X_train, family=binomial)
summary(glm.b)


score <- function(model, X_test, target) {
    prob = predict(model, X_test, type="response")
    pred = rep(1, length(prob))
    pred[prob > 0.5] = 0

    y_pred <- pred
    y_true <- target
    
    print(mean(y_pred==y_true))
}

score(glm.b, eighty_twenty$X_test, eighty_twenty$y_test)


Call:
glm(formula = target ~ income + balance, family = binomial, data = eighty_twenty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.2179  -0.1417  -0.0573  -0.0207   3.7308  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.614900940   0.488362872  -23.78 < 0.0000000000000002 ***
income        0.000021738   0.000005602    3.88             0.000104 ***
balance       0.005677333   0.000254807   22.28 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2340.6  on 7999  degrees of freedom
Residual deviance: 1258.1  on 7997  degrees of freedom
AIC: 1264.1

Number of Fisher Scoring iterations: 8


[1] 0.026


## (c) 
Repeat the process in (b) three times, using three different splits of the observations into a training set and a validation set. Comment on the results obtained:

There is low variance across the three different splits. The standard deviation of the error rates is ~0.003.

In [9]:
sixty_forty = train_test_split(0.6)
glm.c1 = glm(target~income+balance, data=sixty_forty$X_train, family=binomial)
summary(glm.c1)

score(glm.c1, sixty_forty$X_test, sixty_forty$y_test)


Call:
glm(formula = target ~ income + balance, family = binomial, data = sixty_forty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.1187  -0.1542  -0.0615  -0.0232   3.6700  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.146101620   0.531665250  -20.96 < 0.0000000000000002 ***
income        0.000017941   0.000006165    2.91              0.00361 ** 
balance       0.005501280   0.000278994   19.72 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1840.4  on 5999  degrees of freedom
Residual deviance: 1009.9  on 5997  degrees of freedom
AIC: 1015.9

Number of Fisher Scoring iterations: 8


[1] 0.02325


In [10]:
fifty_fifty = train_test_split(0.5)
glm.c2 = glm(target~income+balance, data=fifty_fifty$X_train, family=binomial)
summary(glm.c2)

score(glm.c2, fifty_fifty$X_test, fifty_fifty$y_test)


Call:
glm(formula = target ~ income + balance, family = binomial, data = fifty_fifty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.8274  -0.1345  -0.0525  -0.0188   3.7704  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.831002703   0.667949490 -17.712 < 0.0000000000000002 ***
income        0.000019471   0.000007526   2.587              0.00968 ** 
balance       0.005876903   0.000350610  16.762 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1347.42  on 4999  degrees of freedom
Residual deviance:  737.22  on 4997  degrees of freedom
AIC: 743.22

Number of Fisher Scoring iterations: 8


[1] 0.0282


In [11]:
seventy_thirty = train_test_split(0.7)
glm.c3 = glm(default~income+balance, data=seventy_thirty$X_train, family=binomial)
summary(glm.c3)

score(glm.c3, seventy_thirty$X_test, seventy_thirty$y_test)


Call:
glm(formula = default ~ income + balance, family = binomial, 
    data = seventy_thirty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.5241  -0.1421  -0.0557  -0.0200   3.4817  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.730534830   0.530341966 -22.119 < 0.0000000000000002 ***
income        0.000023229   0.000006112   3.801             0.000144 ***
balance       0.005732021   0.000273938  20.924 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2077.4  on 6999  degrees of freedom
Residual deviance: 1102.1  on 6997  degrees of freedom
AIC: 1108.1

Number of Fisher Scoring iterations: 8


[1] 0.02666667


In [12]:
 sd(c(0.02325, 0.0282, 0.02666667))

[1] 0.00253401

In [13]:
mean(c(0.02325, 0.0282, 0.02666667))

[1] 0.02603889

## (d) 
Now consider a logistic regression model that predicts the probability of default using income, balance, and a dummy variable for student. Estimate the test error for this model using the validation set approach. Comment on whether or not including a dummy variable for student leads to a reduction in the test error rate.

Although income and student may less significate, the standard deviation of the error rates has dropped slightly. The mean error has also dropped. Accordingly, we can derive that accuracy and variance have improved.

In [14]:
for (s in c(123, 345, 456)) {
    set.seed(s)
    seventy_thirty = train_test_split(0.7)

    glm.c2 = glm(target~income+balance+student_e, data=seventy_thirty$X_train, family=binomial)
    print(summary(glm.c2))

    score(glm.c2, seventy_thirty$X_test, seventy_thirty$y_test) 
}


Call:
glm(formula = target ~ income + balance + student_e, family = binomial, 
    data = seventy_thirty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.1583  -0.1409  -0.0567  -0.0205   3.7368  

Coefficients:
                 Estimate    Std. Error z value            Pr(>|z|)    
(Intercept) -11.088952122   0.592275207 -18.723 <0.0000000000000002 ***
income        0.000007592   0.000009799   0.775              0.4385    
balance       0.005759887   0.000277226  20.777 <0.0000000000000002 ***
student_e    -0.522166454   0.282758504  -1.847              0.0648 .  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2043.8  on 6999  degrees of freedom
Residual deviance: 1094.9  on 6996  degrees of freedom
AIC: 1102.9

Number of Fisher Scoring iterations: 8

[1] 0.027

Call:
glm(formula = target ~ income + balance + student_e, family = binomial, 
    data = sevent

In [15]:
sd(c(0.027, 0.02733333, 0.02533333))

[1] 0.001071517

In [16]:
mean(c(0.027, 0.02733333, 0.02533333))

[1] 0.02655555


# 6. 

We continue to consider the use of a logistic regression model to
predict the probability of default using income and balance on the
Default data set. In particular, we will now compute estimates for
the standard errors of the income and balance logistic regression coefficients
in two different ways: 
- (1) using the bootstrap, and 
- (2) using the standard formula for computing the standard errors in the glm() function. Do not forget to set a random seed before beginning your analysis.

In [17]:
seventy_thirty = train_test_split(0.7)
set.seed(1234)
glm.6 = glm(default~income+balance, data=seventy_thirty$X_train, family=binomial)
summary(glm.6)

score(glm.c3, seventy_thirty$X_test, seventy_thirty$y_test)


Call:
glm(formula = default ~ income + balance, family = binomial, 
    data = seventy_thirty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.1881  -0.1415  -0.0565  -0.0208   3.7271  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.584189307   0.525108677  -22.06 < 0.0000000000000002 ***
income        0.000022286   0.000006073    3.67             0.000243 ***
balance       0.005628427   0.000272293   20.67 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2010.0  on 6999  degrees of freedom
Residual deviance: 1088.5  on 6997  degrees of freedom
AIC: 1094.5

Number of Fisher Scoring iterations: 8


[1] 0.02766667


## (a) 

Using the summary() and glm() functions, determine the estimated
standard errors for the coefficients associated with income
and balance in a multiple logistic regression model that uses
both predictors.

In [18]:
summary(glm.6)


Call:
glm(formula = default ~ income + balance, family = binomial, 
    data = seventy_thirty$X_train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.1881  -0.1415  -0.0565  -0.0208   3.7271  

Coefficients:
                 Estimate    Std. Error z value             Pr(>|z|)    
(Intercept) -11.584189307   0.525108677  -22.06 < 0.0000000000000002 ***
income        0.000022286   0.000006073    3.67             0.000243 ***
balance       0.005628427   0.000272293   20.67 < 0.0000000000000002 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2010.0  on 6999  degrees of freedom
Residual deviance: 1088.5  on 6997  degrees of freedom
AIC: 1094.5

Number of Fisher Scoring iterations: 8


## (b) 

Write a function, boot.fn(), that takes as input the Default data
set as well as an index of the observations, and that outputs
the coefficient estimates for income and balance in the multiple
logistic regression model.

In [19]:
library(boot)

In [20]:
boot.fn = function(data, index) {
    
    def = data$target[index]
    inc = data$income[index]
    bal = data$balance[index]
    
    glm.6b = glm(def~inc+bal, family='binomial')
    
    # print(summary(glm.6b))
    return (summary(glm.6b)$coefficients[2:3,2])
}

## (c) 

Use the boot() function together with your boot.fn() function to
estimate the standard errors of the logistic regression coefficients
for income and balance.

In [21]:
samp = seventy_thirty$X_train
boot(samp, boot.fn, 30)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = samp, statistic = boot.fn, R = 30)


Bootstrap Statistics :
         original            bias        std. error
t1* 0.00000607322 0.000000005487568 0.0000002002319
t2* 0.00027229264 0.000001682168558 0.0000120662301

## (d) 

Comment on the estimated standard errors obtained using the
glm() function and using your bootstrap function:

The standard errors for both attributes has dropped significantly after applying the bootstrap.


## 7. 

In Sections 5.3.2 and 5.3.3, we saw that the cv.glm() function can be
used in order to compute the LOOCV test error estimate. Alternatively,
one could compute those quantities using just the glm() and
predict.glm() functions, and a for loop. You will now take this approach
in order to compute the LOOCV error for a simple logistic
regression model on the Weekly data set. Recall that in the context
of classification problems, the LOOCV error is given in (5.4).

In [29]:
data('Weekly')

head(Weekly)

Year Lag1   Lag2   Lag3   Lag4   Lag5   Volume    Today  Direction
1 1990  0.816  1.572 -3.936 -0.229 -3.484 0.1549760 -0.270 Down     
2 1990 -0.270  0.816  1.572 -3.936 -0.229 0.1485740 -2.576 Down     
3 1990 -2.576 -0.270  0.816  1.572 -3.936 0.1598375  3.514 Up       
4 1990  3.514 -2.576 -0.270  0.816  1.572 0.1616300  0.712 Up       
5 1990  0.712  3.514 -2.576 -0.270  0.816 0.1537280  1.178 Up       
6 1990  1.178  0.712  3.514 -2.576 -0.270 0.1544440 -1.372 Down

## (a) 

Fit a logistic regression model that predicts Direction using Lag1 and Lag2.

In [30]:
set.seed(789)

glm.7 = glm(Direction ~ Lag1+Lag2, data=Weekly, family=binomial)
summary(glm.7)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly)

Deviance Residuals: 
   Min      1Q  Median      3Q     Max  
-1.623  -1.261   1.001   1.083   1.506  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22122    0.06147   3.599 0.000319 ***
Lag1        -0.03872    0.02622  -1.477 0.139672    
Lag2         0.06025    0.02655   2.270 0.023232 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1496.2  on 1088  degrees of freedom
Residual deviance: 1488.2  on 1086  degrees of freedom
AIC: 1494.2

Number of Fisher Scoring iterations: 4


## (b) 

Fit a logistic regression model that predicts Direction using Lag1 and Lag2 using all but the first observation.

In [35]:
glm.7b <- glm(Direction ~ Lag1+Lag2, data=Weekly[-c(1),], family=binomial)
summary(glm.7b)


Call:
glm(formula = Direction ~ Lag1 + Lag2, family = binomial, data = Weekly[-c(1), 
    ])

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-1.6258  -1.2617   0.9999   1.0819   1.5071  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.22324    0.06150   3.630 0.000283 ***
Lag1        -0.03843    0.02622  -1.466 0.142683    
Lag2         0.06085    0.02656   2.291 0.021971 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1494.6  on 1087  degrees of freedom
Residual deviance: 1486.5  on 1085  degrees of freedom
AIC: 1492.5

Number of Fisher Scoring iterations: 4


## (c) 

Use the model from (b) to predict the direction of the first observation.
You can do this by predicting that the first observation
will go up if P(Direction="Up"|Lag1, Lag2) > 0.5. Was this observation
correctly classified?

The observation is not correctly classified.

In [42]:
glm.7c = predict(glm.7b, Weekly[1,], type="response")
if(glm.7c > 0.5) "Up" else "Down"
    
Weekly[1,]$Direction

[1] "Up"

[1] Down
Levels: Down Up

## (d) 

Write a for loop from i = 1 to i = n, where n is the number of
observations in the data set, that performs each of the following
steps:

- i. Fit a logistic regression model using all but the ith observation
to predict Direction using Lag1 and Lag2.

- ii. Compute the posterior probability of the market moving up
for the ith observation.

- iii. Use the posterior probability for the ith observation in order
to predict whether or not the market moves up.

In [58]:
preds = rep('Up', length(Weekly$Direction))
err = rep(1, length(Weekly$Direction))

for (i in 1:length(Weekly$Direction)) { 
    model = glm(Direction ~ Lag1+Lag2, data=Weekly[-c(i),], family=binomial)
    probs = predict(model, Weekly[i,], type="response")
  
    preds[i] <- ifelse((probs > 0.5), 'Up', 'Down')    
    err[i] <- ifelse(preds[i]==Weekly$Direction[i], 0, 1)
}

mean(err)

[1] 0.4499541

### iv. 

Determine whether or not an error was made in predicting
the direction for the ith observation. If an error was made,
then indicate this as a 1, and otherwise indicate it as a 0.
(e) Take the average of the n numbers obtained in (d)iv in order to
obtain the LOOCV estimate for the test error. 

Comment on the results:

The LOOCV estimate for the test error is ~45% which tells us that predictions are correct only 55% of the time. We can conclude that model estimates are only moderately better than chance.